staging/src/kubevirt.io/client-go/apis/flavor/v1alpha1/types.go

```diff
@@ -0,0 +1,100 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1 "kubevirt.io/client-go/api/v1"
+)
+
+// VirtualMachineFlavor resource contains common VirtualMachine configuration
+// that can be used by multiple VirtualMachine resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+type VirtualMachineFlavor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualMachineFlavorSpec `json:"spec" valid:"required"`
+}
+
+// VirtualMachineFlavorList is a list of VirtualMachineFlavor resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+type VirtualMachineFlavorList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []VirtualMachineFlavor `json:"items"`
+}
+
+// VirtualMachineClusterFlavor is a cluster scoped version of VirtualMachineFlavor resource.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +genclient
+// +genclient:nonNamespaced
+type VirtualMachineClusterFlavor struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec VirtualMachineFlavorSpec `json:"spec" valid:"required"`
+}
+
+// VirtualMachineClusterFlavorList is a list of VirtualMachineClusterFlavor resources.
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+type VirtualMachineClusterFlavorList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []VirtualMachineClusterFlavor `json:"items"`
+}
+
+// +k8s:openapi-gen=true
+type VirtualMachineFlavorSpec struct {
+	// DefaultProfile is a VirtualMachineFlavorProfile that will be applied to a VirtualMachine,
+	// if no custom profile is specified on the VirtualMachine resource.
+	// +optional
+	DefaultProfile *VirtualMachineFlavorProfile `json:"defaultProfile,omitempty"`
```

Comment:
I think the more idiomatic way of mapping these different kind of profiles is a default field in the profile instead of two fields.
So only one listType map called `Profiles` and then validate that only one has `default: true`.
It might be slower to access and validate though, but from a schema perspective it feels more fitting.

Or the DefaultProfile field could just be a name reference to a profile in the list.
Otherwise, would we have the default profile duplicated so users can specify it by name to be explicit?

I'm not locked on either of the options yet, but on first sight this way of specifying the default doesn't seem right to me.

Reply:
Makes sense. I will change it to use `default: true` field in the profiles. It is better than having a name reference.

Reply:
Done